rust/tools/authorization-logic/README.md

```diff
@@ -0,0 +1,209 @@
+```
+"Alice" says somePred(x) :- condition(x).
+```
+we can combine this with the fact `"Alice" says condition(x)` to prove `"Alice" 
+says somePred(x)`.
+
+When writing facts or rules, the arguments to predicates can either be 
+constants or variables. Constants are written in quotes whereas variables are 
+not. For example,
+```
+"Alice" says somePred("x") :- condition("x").
+```
+gives a rule for proving somePred for just a particular argument, `"x"`, 
+whereas the same rule replaced with `x` can be applied to any argument about 
+which we can prove `"Alice" says condition(x)`.
+
+Principal names can also be used as arguments to predicates, and quotes are used
+indicate whether a principal is a constant or variable in the same way.
+
+Multiple rules/facts stated by the same principal can be grouped together using 
+curly braces, as in:
+```
+"Alice" says {
+    rule1(x) :- condition(x).
+    rule2(y) :- condition1(y), condition(x).
+    someFact(x).
+}
+```
+
+#### Delegation
+
+Delegation works by prepending a predicate on the left of a fact or rule with 
+`<principal_name> canSay`. Delegations can only appear on the left side of a 
+rule. As an example, the rule
+```
+"Alice" says "Bob" canSay pred(x) :- condition(x).
+```
+can be combined with `"Alice" says condition("y")` and `"Bob" says pred("y")` 
+to prove `Alice says pred("y")`.
+
+#### Attributes and canActAs
+
+Predicates can be used to represent attributes of principals or other 
+constants/variables using the syntax `principal_name predicate_name(<args>)`, 
+such as, `"Manager" hasTraining("interviewing")`. Attributes can appear in 
+facts or in the left or right of a rule. 
+
+Attributes can be passed from one principal to another using `canActAs`. If `A 
+canActAs B`, then `A` inherits all the attributes of `B`. For example,
+```
+"HR" says "Alice" canActAs "Manager".
+```
+can be combined with `"HR" says "Manager" hasTraining("interviewing")` to prove
+`"HR" says "Alice" hasTraining("interviewing")`. 
+
+Combining attributes with `canActAs` gives a way to define _roles_ which group 
+privileges together.
+
+#### Queries
+Queries test if it is possible to prove a specific predicate given the facts 
+and rules. Queries can be written using the syntax:
+```
+<query_name> = query <principal_name> says <predicate_name>(<args>)?
+```
+
+If the program contained queries (described below), then the there will also be
+a file called `<query_name>.csv` for each query. The query is true if the CSV
+contains one entry and is false otherwise.
+
+#### Importing and Exporting statements
+
+Statements can be imported/exported from potentially untrusted sources such as 
+remote machines or storage. To authenticate statements, they are digitally 
+signed using ECDSA signatures when exporting, and the signatures are checked 
+when importing. 
+
+To support exporting, a principal must be bound to a private key. Private and 
+public keys must be serialized in JSON. See 
```

Comment:
I couldn't find a JSON schema, but for the ECDSA keys it would use the JSON equivalent of the following protobuf schema (https://github.com/project-oak/tink-rust/blob/main/proto/proto/ecdsa.proto).

Reply:
Hmmmmm... neither seem like the ideal reference, but the code that actually handles the JSON from the tink-rust imlementation seems more relevant, so I linked that one. Good enough?